### PR TITLE
doc: kernel/mm: add a page about virtual memory

### PR DESCRIPTION
**Pull request by dcpleung**
Add a page to briefly talk about virtual memory.

### CODE REVIEW

**Review by kartben — CHANGES_REQUESTED** [review]
Very nice, and nice touch with the use of `emphasize-lines` :)
My only actual request for change would be to stick to the heading style specified in the doc guidelines, i.e. equal sign for H3. Thanks!

**Review by andyross — DISMISSED** [review]
Let me throw a +1 on this just to flag that I've seen it and think it should go in.  I really should spend some time on an editting pass, but don't have time in the immediate term.  So if I forget, treat this as my +1.

File: doc/kernel/memory_management/virtual_memory.rst [hunk 1]
```diff
@@ -0,0 +1,199 @@
...
+  * Although the virtual memory space can be larger than physical
+    memory space, without enabling demand paging, all virtually
+    mapped memory must be backed by physical memory.
+
+
+Kconfigs
+********
+
+Required
+^^^^^^^^
```
[the remark this answers is not in the document]
Comment (dcpleung) [reply doc/kernel/memory_management/virtual_memory.rst]: I tried a clean build and it worked now with `=====`. Must be issue with incremental build.

File: doc/kernel/memory_management/virtual_memory.rst [hunk 2]
```diff
@@ -0,0 +1,199 @@
...
+##############
+
+Virtual memory (VM) in Zephyr provides developers the ability to fine tune
+access to memory. To utilize virtual memory, the platform must support
+Memory Management Unit (MMU) and is enabled in the build. Due to
+the target of Zephyr mainly being embedded systems, virtual memory
+support in Zephyr differs a bit from that in traditional operating
+systems:
+
+* **Mapping of Kernel Image**
```
[the remark this answers is not in the document]
Comment (dcpleung) [reply doc/kernel/memory_management/virtual_memory.rst]: Definition list made it look a lot better. Thanks for the suggestion.

File: doc/kernel/memory_management/virtual_memory.rst [hunk 3]
```diff
@@ -0,0 +1,197 @@
...
+* ``Z_FREE_VM_START`` is the beginning of the virtual address space where addresses
+  can be allocated for memory mapping. This depends on whether
+  :kconfig:option:`CONFIG_ARCH_MAPS_ALL_RAM` is enabled.
+
+  * If it is enabled, which means all physical memory are mapped in virtual
+    memory address space, and it is the same as
+    (:kconfig:option:`CONFIG_SRAM_BASE_ADDRESS` + :kconfig:option:`CONFIG_SRAM_SIZE`).
+
+  * If it is disabled, ``Z_FREE_VM_START`` is the same ``Z_KERNEL_VIRT_END`` which
+    is the end of the kenerl image.
```
Comment (kartben) [inline doc/kernel/memory_management/virtual_memory.rst:127]:
```suggestion
  * If it is disabled, ``Z_FREE_VM_START`` is the same as ``Z_KERNEL_VIRT_END``, which
    is the end of the kernel image.
```

File: doc/kernel/memory_management/virtual_memory.rst [hunk 4]
```diff
@@ -0,0 +1,197 @@
+.. _memory_management_api_virtual_memory:
+
+Virtual Memory
+##############
+
+Virtual memory (VM) in Zephyr provides developers the ability to fine tune
```
Comment (kartben) [inline doc/kernel/memory_management/virtual_memory.rst:6]:
```suggestion
Virtual memory (VM) in Zephyr provides developers with the ability to fine tune
```

File: doc/kernel/memory_management/virtual_memory.rst [hunk 5]
```diff
@@ -0,0 +1,197 @@
...
+access to memory. To utilize virtual memory, the platform must support
+Memory Management Unit (MMU) and is enabled in the build. Due to
```
Comment (kartben) [inline doc/kernel/memory_management/virtual_memory.rst:8]:
```suggestion
Memory Management Unit (MMU) and it must be enabled in the build. Due to
```

File: doc/kernel/memory_management/virtual_memory.rst [hunk 6]
```diff
@@ -0,0 +1,197 @@
...
+    mapped memory must be backed by physical memory.
+
+
+Kconfigs
+********
+
+Required
+========
+
+These are the kconfigs that need to be enabled or defined for kernel to support
```
Comment (kartben) [inline doc/kernel/memory_management/virtual_memory.rst:40]:
```suggestion
These are the Kconfigs that need to be enabled or defined for kernel to support
```

File: doc/kernel/memory_management/virtual_memory.rst [hunk 7]
```diff
@@ -0,0 +1,197 @@
...
+  the virtual address space. This is useful for mapping device MMIO regions for
+  more precise access control.
+
+
+Memory Map Overview
+*******************
+
+This is an overview of the memory map of the virtual memory address space.
+Note that the ``Z_*`` macros, which are used in code, may have different
+meanings depending on architecture and kconfigs, which will be explained
```
Comment (kartben) [inline doc/kernel/memory_management/virtual_memory.rst:70]:
```suggestion
meanings depending on architecture and Kconfigs, which will be explained
```

**Review by kartben — CHANGES_REQUESTED** [review]
Brilliant -- thanks!
Just a few very minor typos/spelling corrections here and there.

File: doc/kernel/memory_management/virtual_memory.rst (hunk 4 above)
Comment (dcpleung) [reply doc/kernel/memory_management/virtual_memory.rst]: Fixed.

File: doc/kernel/memory_management/virtual_memory.rst (hunk 5 above)
Comment (dcpleung) [reply doc/kernel/memory_management/virtual_memory.rst]: Fixed.

File: doc/kernel/memory_management/virtual_memory.rst (hunk 7 above)
Comment (dcpleung) [reply doc/kernel/memory_management/virtual_memory.rst]: Fixed.

File: doc/kernel/memory_management/virtual_memory.rst
[the remark this answers is not in the document]
Comment (dcpleung) [reply doc/kernel/memory_management/virtual_memory.rst:1]: Oh... I did not reformat this one to make it easier to review the latest small changes.

**Review by cfriedt — APPROVED** [review]

**Review by kartben — APPROVED** [review]

**Review by nashif — APPROVED** [review]

**Pull request merged**